gpMgmt/bin/gprebalance_modules/rebalance_validator.py

```diff
@@ -0,0 +1,91 @@
+from typing import List
+from gprebalance_modules.rebalance import Host, MirrorStrategy, dbconn, GpArray,  Segment, MODE_NOT_SYNC, STATUS_DOWN
+
+
+class StateValidationError(Exception):
+    pass
+
+
+class ClusterValidator:
+    def __init__(self, hosts: List[Host], segarray: List[Segment], has_mirrors: bool, mirror_strategy: MirrorStrategy):
+        self.hosts = hosts
+        self.mirror_strategy = mirror_strategy
+        self.existing_hosts = [
+            h for h in hosts if h.primary_segments or h.mirror_segments]
+        self.new_hosts = [h for h in hosts if not bool(
+            h.primary_segments) and not bool(h.mirror_segments)]
+        self.segarray = segarray
+        self.has_mirrors = has_mirrors
+
+    def validate_segment_status(self):
+        inv = self.get_invalid_segments()
+        if len(inv) > 0:
+            raise StateValidationError(
+                f"The {[s.content for s in inv]} segments are down")
+
+    def get_invalid_segments(self) -> List[Segment]:
+        l = []
+        for seg in self.segarray:
+            if not seg.valid:
+                l.append(seg)
+        return l
```

Comment:
Self has `existing_hosts` and `new_hosts`. Why not do the same for `invalid_segments`?